File: solutions/javascript/pig-latin/1/pig-latin.js

// We will use regular expressions to match the rules and break down the words in the relevant pieces.
// More general rules (like rule2) will be placed lower to let the more specific rules match first.

function translateWord(word) {

  const rule1 = word.match(/^([aeiou]|xr|yt).*$/)
  if (rule1) return `${rule1[0]}ay`;

  const rule3 = word.match(/^([^aeiou]?)qu(.*)$/);
  if (rule3) return `${rule3[2]}${rule3[1]}quay`;

  const rule4 = word.match(/^([^aeiou]+)y(.*)$/);
  if (rule4) return `y${rule4[2]}${rule4[1]}ay`;

  const rule2 = word.match(/^([^aeiou]+)(.*)$/)
  if (rule2) return `${rule2[2]}${rule2[1]}ay`;

  return word;

};

export function translate(sentence) {
  return sentence.split(' ').map(w => translateWord(w)).join(' ');
}

File: solutions/python/pig-latin/1/pig_latin.py
import re

def translate_word(text):

    # Rule 1, word starts with a vowel.
    if re.match(r'([aeiou]|xr|yt)', text):
        return text + 'ay'

    # We have eliminated words starting with vowel,
    # The word has to start with a consonant.
    # Use this to assume the first letter is a consonant.

    # Rule 3, (single) consonant followed by 'qu'.
    if text[1:].startswith('qu'):
        return text[3:] + text[0:3] + 'ay'

    # Rule 4 derives directly from rule 2 if we correctly
    # identify the consonant clusters.
    # Rule 2, words starts with consonant (including consonant cluster).
        
    # Check for consonent cluster (considering 'qu' as a cluster).
    cluster_size = re.match(r'(qu|([^aeiou][^aeiouy]*))', text).end()
    return text[cluster_size:] + text[0:cluster_size] + 'ay'

    # we should have handled all the cases above
    raise ValueError(f'no rule for {text}')

def translate(text):
    new_words = [translate_word(w) for w in text.split(' ')]
    return ' '.join(new_words)